### ISSUE TITLE
[Question] type() function slow

### ISSUE THREAD

**Issue by dllyd-work** [glitchassassin/lackey#91]
Hi,

I'm trying to convert my program from sikulix into lackey and I've found that it is running 100% slower. Specifically with the type() function.

I ran the same test in sikulx and it ran in 25 seconds while in lackey it took 55 seconds.
I have tried setting the type(value, 0) but it seemed to make no difference.

Thanks,
Daniel

**Comment by dllyd-work** [comment]
It was slow specifically while doing type(Key.ENTER, 0) changing this into keyDown(Key.ENTER) and keyUp(Key.ENTER) speeds it up significantly.

**Comment by glitchassassin** [comment]
Currently Lackey imposes a default delay of 0.05s between each keypress. You can change this by setting `Settings.TypeDelay` to, for example, 0.01:

    Settings.TypeDelay = 0.01
    type("some very long string with lots \n of \n new lines \n in it") # types faster

I'm not sure that's how Sikuli behaves by default, so I'll review it to see if this should be changed. Let me know if the above workaround doesn't work for you in the mean time.

**Comment by nejch** [comment]
I had a look at this because it was affecting my scripts as well. I think there are two issues here, one is that there is a hardcoded `time.sleep(0.2)` in the Region `type()` method. So if you're calling it several times or using it in loops it will be significantly slower than sikuli.

For me, an example would be pressing Down repeatedly in a dropdown to get to an item instead of opening the list. Not sure if I'm missing something but would it maybe be better to have the users themselves provide explicit waits for slow applications? I see other methods use it as well though, so maybe you want to have a look first if they all could be changed.

Removing the wait time gives almost identical times in loops and strings (like something below).
```
    def test_type_a(self):
        lackey.type('aaaaaaaaaaaaaaaaaaaa')

    def test_type_a_loop(self):
        for _ in range(20):
            lackey.type('a')

```

The second issue is how lackey internally parses special codes. It waits on every iteration/character when inside `{}` (noticeable with longer codes like PAGE_DOWN), even when it ends up only typing one special key. E.g.:
```
import cProfile
import lackey

cProfile.run('lackey.type("")', sort='tottime')
cProfile.run('lackey.type("a")', sort='tottime')
cProfile.run('lackey.type("{CTRL}")', sort='tottime')
cProfile.run('lackey.type("{PAGE_DOWN}")', sort='tottime')
cProfile.run('lackey.type("{NOT_A_SPECIAL_CODE}")', sort='tottime')
```

With the default `Settings.TypeDelay `this gives you (truncated here for clarity):
```
[action] Typing '' with modifiers 'None'
         16 function calls in 0.199 seconds

[action] Typing 'a' with modifiers 'None'
         81 function calls in 0.263 seconds

[action] Typing '{CTRL}' with modifiers 'None'
         71 function calls in 0.595 seconds

[action] Typing '{PAGE_DOWN}' with modifiers 'None'
         81 function calls in 0.893 seconds

[action] Typing '{NOT_A_SPECIAL_CODE}' with modifiers 'None'
         2160 function calls (2158 primitive calls) in 3.521 seconds
```
So `time.sleep()` is called 12 times for Page Down and takes almost a second. With `if delay and not in_special_code:` it looks much better:
```
[action] Typing '{PAGE_DOWN}' with modifiers 'None'
         71 function calls in 0.256 seconds
```

**Comment by glitchassassin** [comment]
Thanks for the analysis! I'll run my extended test cases to make sure this doesn't break anything and then merge your pull request.

**Comment by nejch** [comment]
Thanks! :)

**Comment by glitchassassin** [comment]
This issue has been fixed with v0.7.0.
